b9/src/ExecutionContext.cpp

```diff
@@ -133,96 +148,150 @@ StackElement ExecutionContext::interpret(const std::size_t functionIndex) {
         doIntDiv();
         break;
       case OpCode::INT_PUSH_CONSTANT:
-        doIntPushConstant(instructionPointer->immediate());
+        doIntPushConstant();
         break;
       case OpCode::INT_NOT:
         doIntNot();
         break;
       case OpCode::INT_JMP_EQ:
-        instructionPointer += doIntJmpEq(instructionPointer->immediate());
+        doIntJmpEq();
         break;
       case OpCode::INT_JMP_NEQ:
-        instructionPointer += doIntJmpNeq(instructionPointer->immediate());
+        doIntJmpNeq();
         break;
       case OpCode::INT_JMP_GT:
-        instructionPointer += doIntJmpGt(instructionPointer->immediate());
+        doIntJmpGt();
         break;
       case OpCode::INT_JMP_GE:
-        instructionPointer += doIntJmpGe(instructionPointer->immediate());
+        doIntJmpGe();
         break;
       case OpCode::INT_JMP_LT:
-        instructionPointer += doIntJmpLt(instructionPointer->immediate());
+        doIntJmpLt();
         break;
       case OpCode::INT_JMP_LE:
-        instructionPointer += doIntJmpLe(instructionPointer->immediate());
+        doIntJmpLe();
         break;
       case OpCode::STR_PUSH_CONSTANT:
-        doStrPushConstant(instructionPointer->immediate());
-        break;
-      case OpCode::STR_JMP_EQ:
-        // TODO
-        break;
-      case OpCode::STR_JMP_NEQ:
-        // TODO
+        doStrPushConstant();
         break;
       case OpCode::NEW_OBJECT:
         doNewObject();
         break;
       case OpCode::PUSH_FROM_OBJECT:
-        doPushFromObject(Om::Id(instructionPointer->immediate()));
+        doPushFromObject();
         break;
       case OpCode::POP_INTO_OBJECT:
-        doPopIntoObject(Om::Id(instructionPointer->immediate()));
+        doPopIntoObject();
         break;
       case OpCode::CALL_INDIRECT:
         doCallIndirect();
         break;
       case OpCode::SYSTEM_COLLECT:
         doSystemCollect();
         break;
+      case OpCode::END_SECTION:
+        continue_ = false;
+        break;
```

Comment:
I'm assuming this is to gracefully handle the case where a function has no explicit return instruction?

Reply:
Yes. This shouldn't normally happen though.

Since moving the stack frames to the interpreter stack, I needed a way to halt execution in the VM. when the main function returns, the interpreter will try to resume the outermost call frame, and we need to pause execution.

One idea, was to wrap the first call in a pseudo function body, which when resumed, will halt the interpreter. The body of the pseudo function would be something like:

```lisp
(push-param 0)
(call-indirect)
(end-section)
```

The other strategy, which is probably more obvious, is to tag the outermost frame. When `(function-return)` resumes the outermost frame, it stops execution.

I am leaning towards using an outermost pseudo function, but still considering the options.